Description: Better protection of UNIX socket names.
 Ensure that a NUL character is always present in the socket name.
 Use SUN_LEN to calculate name lengths for optimal portability.
Author: Mats Erik Andersson <debian@gisladisker.se>
Last-Update: 2011-11-11

diff -Naurp ttysnoop-0.12d.debian/ttysnoop.c ttysnoop-0.12d/ttysnoop.c
--- ttysnoop-0.12d.debian/ttysnoop.c
+++ ttysnoop-0.12d/ttysnoop.c
@@ -10,6 +10,8 @@
 	v0.12	6-9-94	     ""		- see ttysnoops.c
 */
 
+#define _BSD_SOURCE	/* SUN_LEN from <sys/un.h> */
+
 #include <sys/types.h>
 #include <sys/stat.h>
 #include <sys/ioctl.h>
@@ -60,9 +62,9 @@ int main (int argc, char *argv[])
 	snprintf (sockname, sizeof(sockname), "%s/%s", SPOOLDIR, argv[1]);
 	memset (&sock_addr, '\0', sizeof(sock_addr));
 	sock_addr.sun_family = AF_UNIX;
-	strncopy (sock_addr.sun_path, sockname);
+	strncpy (sock_addr.sun_path, sockname, sizeof(sock_addr.sun_path) - 1);
 
-	if (connect(sockfd, (struct sockaddr *) &sock_addr, sizeof(sock_addr)) < 0)
+	if (connect(sockfd, (struct sockaddr *) &sock_addr, SUN_LEN(&sock_addr)) < 0)
 		errorf ("Cannot connect to server. Is it really a valid socket?\n");
 	
 	/* put stdin into raw mode */
diff -Naurp ttysnoop-0.12d.debian/ttysnoops.c ttysnoop-0.12d/ttysnoops.c
--- ttysnoop-0.12d.debian/ttysnoops.c
+++ ttysnoop-0.12d/ttysnoops.c
@@ -19,6 +19,8 @@
 */
 
 #define _XOPEN_SOURCE 600 /* ptsname(), posix_openpt() */
+#define _BSD_SOURCE   /* SUN_LEN from <sys/un.h> */
+
 #include <sys/types.h>
 #include <sys/stat.h>
 #include <sys/ioctl.h>
@@ -578,8 +580,8 @@ int main (int argc, char *argv[])
 		unlink (sockname);
 		memset(&serv_addr, '\0', sizeof(serv_addr));
 		serv_addr.sun_family = AF_UNIX;
-		strncopy (serv_addr.sun_path, sockname);
-		if (bind(servfd, (struct sockaddr *) &serv_addr, sizeof(serv_addr)) < 0)
+		strncpy (serv_addr.sun_path, sockname, sizeof(serv_addr.sun_path) - 1);
+		if (bind(servfd, (struct sockaddr *) &serv_addr, SUN_LEN(&serv_addr)) < 0)
 			errorf ("can't bind server socket (%s)\n", sockname);
 			
 		if (listen(servfd, 5) < 0)
